### PR TITLE
Remove legacy code of single user device resync api

### LINKED ISSUE #14808: Remove legacy singular replication endpoint for resyncing remote user devices
I.e. all the code deprecated in #14716, this is pretty much:

* `user_device_resync` methods (I think? I don't see a reason to have these that just call the `multi_` version)
* `ReplicationUserDevicesResyncRestServlet`

I think that's it. 👍

### PR DESCRIPTION
**Pull request by akstron**
Closes #14808

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

Signed-off-by: Alok Kumar Singh alokaks601@gmail.com

### CODE REVIEW

File: changelog.d/15418.removal
Comment (DMRobertson) [inline changelog.d/15418.removal:1]: Note that #14932 is doing the same thing here.

File: changelog.d/15418.removal [hunk 1]
```diff
@@ -0,0 +1 @@
+Since multi-user device resync api is now available, single user counterpart is removed.
```
Comment (DMRobertson) [inline changelog.d/15418.removal:1]: While this is a removal, it is a removal of an internal-only implementation detail: please rename this to `.misc`.

```suggestion
Always use multi-user device resync replication endpoints.
```

**Comment by DMRobertson** [comment]
Note that #14932 makes the same tidy-up---though it seems to remove much more?

File: changelog.d/15418.removal (hunk 1 above)
Comment (akstron) [reply changelog.d/15418.removal]: Updated it in recent commit.

**Review by clokep — CHANGES_REQUESTED** [review]
As @DMRobertson pointed out -- there's quite a bit more to do here.

At the very least we should be able to remove the actual `ReplicationUserDevicesResyncRestServlet` class.

**Comment by akstron** [comment]
Updated the code to remove any usage of single-user resync endpoint.

**Comment by clokep** [comment]
We're getting there! 🎉 I think we also should remove each of the `user_device_resync` implementations. Hopefully there are no more callers of it?

**Comment by akstron** [comment]
Removed each of `user_device_resync` implementation and replaced few calls which were using it with the multi-user counterpart.

File: synapse/handlers/device.py [hunk 2]
```diff
@@ -948,35 +943,8 @@ async def multi_user_device_resync(
...
-        Returns:
-            A dict with device info as under the "devices" in the result of this
-            request:
-            https://matrix.org/docs/spec/server_server/r0.1.2#get-matrix-federation-v1-user-devices-userid
-            None when we weren't able to fetch the device info for some reason,
-            e.g. due to a connection problem.
-        """
-        return (await self.multi_user_device_resync([user_id]))[user_id]
+        except:
+            raise
```
Comment (DMRobertson) [inline synapse/handlers/device.py:947]: An except that immediately raises (without e.g. extra logging or a fallback) doesn't do anything.

I think this should just be

```python
        return await self._multi_user_device_resync_client(user_ids=user_ids)
```

**Review by DMRobertson — CHANGES_REQUESTED** [review]
Thanks, this Looks good!  I'll let @clokep double check once he's back.

One small nit:

File: synapse/handlers/device.py (hunk 2 above)
Comment (akstron) [reply synapse/handlers/device.py]: Got it! Updated it in recent commit.

File: synapse/handlers/device.py [hunk 3]
```diff
@@ -1196,10 +1161,9 @@ async def _maybe_retry_device_resync(self) -> None:
             for user_id in need_resync:
                 try:
                     # Try to resync the current user's devices list.
```
Comment (clokep) [inline synapse/handlers/device.py:1163]: We should probably add a TODO comment here that says to resync multiple users at once...

File: synapse/handlers/federation_event.py [hunk 4]
```diff
@@ -167,8 +169,8 @@ def __init__(self, hs: "HomeServer"):
 
         self._send_events = ReplicationFederationSendEventsRestServlet.make_client(hs)
         if hs.config.worker.worker_app:
-            self._user_device_resync = (
-                ReplicationUserDevicesResyncRestServlet.make_client(hs)
+            self._multi_user_device_resync = (
+                ReplicationMultiUserDevicesResyncRestServlet.make_client(hs)
```
Comment (clokep) [inline synapse/handlers/federation_event.py:173]: I was going to suggest renaming all of these things to remove the `multi`, that can be done as a follow-up though.

**Review by DMRobertson — APPROVED** [review]
Thank you for seeing this through.

**Pull request merged**